### ISSUE TITLE
FileWriter Builder Pattern

### ISSUE THREAD

**Issue by darkfrog26** [outr/scribe#73]
In 2.5 basic builder pattern support was added, but for the more complex scenarios, it is a pain to configure manually.  We need to support clean setup of the `FileWriter`.

**Comment by darkfrog26** [comment]
@cornerman, I've been trying to come up with a good DSL pattern for creating a file writer with all the options (gzip, renaming, etc.) and would be interested in any thoughts you might have here.

**Comment by darkfrog26** [comment]
Not exactly, but you can take a look at this: https://github.com/outr/scribe/blob/master/core/jvm/src/main/scala/scribe/writer/file/LogFileManager.scala

It should give you a glimpse at what I'm trying to accomplish.

**Comment by darkfrog26** [comment]
Feature list:
* Maximum number of logs (delete old logs)
* Maximum sized log files
  * Roll-off old records support
  * Create additional files
* Renaming logs (primarily for standardized name for current log file being renamed to a date stamp)
* GZip support

**Comment by darkfrog26** [comment]
Scenarios:
* Rolling log files (app.log renames to app-YYYY-MM-DD.log after each day)
  * `FileWriter.nio.path("log/app.log").rolling(FileWriter.format.daily)`
* GZip old log files
  * `FileWriter.io.path(FileWriter.format.daily).gzip()`

**Comment by andreak** [comment]
Looks good.
Any reason you've switched to having Path+fileName in one String, or is just for the example showing the DSL-pattern?

**Comment by darkfrog26** [comment]
@andreak, the `path` method looks like: `def path(logPath: => Path)` allowing you to define any mechanism you want for deriving a path.  I'm working on a new prototype right now that may offer even greater flexibility.

**Comment by andreak** [comment]
I see. Works for me.

**Comment by cornerman** [comment]
@darkfrog26 The scenarios look nice! One question: What type does `FileWriter.format.daily` have? It might make sense to move these format functions to another object, because otherwise you might confuse `FileWriter.format` with a factory for a FileWriter. It should be clear, that there are two options for constructing a filewriter: either `io` or `nio`.

> Roll-off old records support

Is a record one logline?

> Create additional files

What kind of file can we create additionally? Is this having a composed writer, which writes to multiple `Writer`?

> Renaming logs (primarily for standardized name for current log file being renamed to a date stamp)

On what occasions would such a renaming occur? Is this separate operation or more like the building block of formatting a filename?

**Comment by cornerman** [comment]
What comes to my mind in the first example:
> FileWriter.nio.path("log/app.log").rolling(FileWriter.format.daily)

Is it necessary to provide a `path` in this example or could we directly use the name resulting from the `rolling`-format? The format can often already identify which one is the latest file.

**Comment by darkfrog26** [comment]
@cornerman, great questions!

The daily format looks like this:
```scala
    lazy val daily: Long => String = (l: Long) => {
      p"${l.t.Y}-${l.t.m}-${l.t.d}"
    }
```

I agree, I'll end up doing a lot of cleanup work after I have this thing working to the extent it can actually support all the scenarios I'm trying to support.

The `io` and `nio` are modifiers, so you can exclude them and it will use the default (which will eventually be determined automatically choosing the optimal solution).

A record may be several log lines.  For example, logging of a `Throwable` would potentially be many many lines.  As well, 2.6 will support varargs for values being logged and they will be appended (in whatever format is used for output) as one log record with potentially several lines.

"Create additional files" this is meant to support the concept of "app.log" getting too big and it rolling it to "app-1.log" and starting a new "app.log" file for example.

"Renaming logs", yes, it's primarily focused toward rolling logs, but hopefully flexible enough to support other potential scenarios.

All options are optional and I'm creating a paradigm with `Action`s that should make it so you can just do `FileWriter.rolling(daily)` and it just works...hopefully.

**Comment by cornerman** [comment]
Thanks for explanation, sounds good to me! 

About `io` and `nio`: How do you actually decide which one is more fitting? :)

**Comment by darkfrog26** [comment]
Most of the time `io` is faster, but if I recall, on MacOS `nio` is faster.  I may need to write a detection of the FS it's running on or something.  That's probably a 2.7 feature. ;)

**Comment by darkfrog26** [comment]
Quick update regarding rolling files:

`FileWriter().path(Paths.get("logs/test.log")).rolling(LogFilePaths.daily("logs/test-", ".log"))`
Will create `logs/test.log` and write to it for a day. When a new day starts it will rename `logs/test.log` to `logs/test-YYYY.MM.DD.log` and create a new `logs/test.log`.

Removing `path(Paths.get("logs/test.log"))` will just make it use the default `logs/app.log` starting path.  If you want to do daily log files so it always creates `logs/test-YYYY.MM.DD.log` you would just do:

`FileWriter().path(LogFilePaths.daily("logs/test-", ".log"))`

Is that acceptable behavior?

**Comment by andreak** [comment]
I find it a bit strange that the default-filename must include the "separator-char" '-'. Makes more sense having it as a separate argument/variable somewhere, maybe with a default value of '-'.
It would be nice to be able to place the date-pattern elsewhere.
Some might want the ```YYYY.MM.DD``` last in the name, like ```"logs/test.log-YYYY.MM.DD"``` or ```"logs/test.log.YYYY.MM.DD"```

**Comment by darkfrog26** [comment]
@andreak, good points and the `LogFilePaths` is just boilerplate right now as I haven't gotten to that part yet.  Are you okay with the combination of `path` and `rolling`?  Is that easily understandable?

**Comment by andreak** [comment]
Yes, and I assume one can append .gzip()? as in ```FileWriter().path(Paths.get("logs/test.log")).rolling(LogFilePaths.daily("logs/test-", ".log")).gzip()```

**Comment by darkfrog26** [comment]
Actually, because of the nature of how and where GZip must be applied, it's `rolling(path, gzip = true)`. The same for `path(path, gzip = true)`. When a file is rolled off or a path change is detected, the old file will be gzipped.

**Comment by darkfrog26** [comment]
Any other scenarios you guys can think of that need to be supported?

**Comment by darkfrog26** [comment]
@andreak and @cornerman, take a look at the current implementation when you have a chance: https://github.com/outr/scribe/blob/2.6/core/jvm/src/main/scala/scribe/writer/FileWriter.scala

It's not finished, but most of the functionality is there. Let me know if you see any mistakes or missing functionality.

**Comment by andreak** [comment]
is ```checkRate``` for having Scribe check every ```rate``` millis if it's necessary to roll the file, and not for every log-record?

**Comment by andreak** [comment]
BTW; The project now builds in IntelliJ by only opening the build.sbt file. No extra steps needed yej!

**Comment by darkfrog26** [comment]
@andreak, yes, `checkRate` is just a performance optimization when logging in a tight loop to avoid unnecessary CPU time wasting.

Did IntelliJ not build for you before?  It always has for me...

**Comment by andreak** [comment]
IntelliJ barfed on some red code and it seemed it didn't understand project-deps correctly. I didn't run any ```gen-idea``` sbt-command. I don't do now either, but now it works out of the box.

**Comment by cornerman** [comment]
This looks really good to me so far! Looking forward to see the implementation :)

When we want to have a maximum number of logs, can we have a possibility to invoke different actions (not only delete but also something like archive). Then one could keep a number of logs on the host and archive older logs to another storage medium. That might even be interesting for the rolling filewriter, where we could role into another storage. Does that make sense?

**Comment by darkfrog26** [comment]
@cornerman, that's a really good idea...give me a few to try to incorporate that.  I already support this with `maxSize` for maximum log file size, but it shouldn't be difficult for to incorporate the same concept for `maxLogs`.  I'll get back to you. :)

**Comment by darkfrog26** [comment]
How about a `logManager` function:
```scala
def maxLogs(max: Int,
              lister: Path => List[Path] = MaxLogFilesAction.MatchLogAndGZInSameDirectory,
              logManager: Path => Unit = Files.deleteIfExists(_),
              checkRate: Long = 10L): FileWriter = ...
```

By default, as you can see it will delete old logs, but you can override this functionality to do whatever you want with those files.

**Comment by cornerman** [comment]
Nice, looks good to me!

**Comment by cornerman** [comment]
Can we add the unit to the `checkRate`?
